### PR TITLE
fix memory leak in swift::sys::Cache

### PR DESCRIPTION
**Pull request by marcrasi**
`swift::sys::Cache` fails to destroy the key one case, causing a small memory leak in SourceKit.

(Caught while running SourceKit under https://gperftools.github.io/gperftools/heap_checker.html)

### CODE REVIEW

File: lib/Basic/Cache.cpp [hunk 1]
```diff
@@ -78,8 +78,10 @@ void CacheImpl::setAndRetain(void *Key, void *Value, size_t Cost) {
   DefaultCacheKey CKey(Key, &DCache.CBs);
   auto Entry = DCache.Entries.find(CKey);
   if (Entry != DCache.Entries.end()) {
-    if (Entry->second == Value)
+    if (Entry->second == Value) {
+      DCache.CBs.keyDestroyCB(Key, nullptr);
```
Comment (benlangmuir) [inline lib/Basic/Cache.cpp]: The documentation says we should release the original key, which is also how it behaves on macOS.  I think we should swap `Entry->first.Key` for the new key here.

Comment (marcrasi) [reply lib/Basic/Cache.cpp]: Done.

File: lib/Basic/Cache.cpp [hunk 2]
```diff
@@ -78,14 +78,16 @@ void CacheImpl::setAndRetain(void *Key, void *Value, size_t Cost) {
...
-      return;
     DCache.CBs.keyDestroyCB(Entry->first.Key, nullptr);
-    DCache.CBs.valueReleaseCB(Entry->second, nullptr);
+    if (Entry->second != Value) {
+      DCache.CBs.valueReleaseCB(Entry->second, nullptr);
+      DCache.CBs.valueRetainCB(Value, nullptr);
+    }
     DCache.Entries.erase(Entry);
+  } else {
+    DCache.CBs.valueRetainCB(Value, nullptr);
```
Comment (benlangmuir) [inline lib/Basic/Cache.cpp]: This logic is getting kind of hard to follow, and I'm not sure how safe it is to `keyDestroy` but leave the key in the map temporarily and then call `operator[]` which will compare against that dead key.  How about something like this (untested):

```c++
if (Entry == DCache.Entries.end()) {
  DCache.CBs.valueRetainCB(Value, nullptr);		
  DCache.Entries[CKey] = Value;
  return;
}

// Destroy the original key, which must be == the new key.
std::swap(Entry->first.Key, Key);
DCache.CBs.keyDestroyCB(Key);

// Replace the value, if necessary.
if (Entry->second != Value) {
  DCache.CBs.valueRetainCB(Value, nullptr);	
  std::swap(Entry->second, Value);
  DCache.CBs.valueReleaseCB(Value, nullptr);
}
```

Comment (marcrasi) [reply lib/Basic/Cache.cpp]: Thanks, that's a lot clearer and it works.

**Review by benlangmuir — APPROVED** [review]

**Comment by benlangmuir** [comment]
@swift-ci please test

**Comment by benlangmuir** [comment]
@swift-ci please test macOS

**Comment by jckarter** [comment]
Could this be causing the ASAN failure we're seeing on the bots?

https://ci.swift.org/job/oss-swift-incremental-ASAN-RA-osx/3572/consoleText

```
******************** TEST 'Swift-Unit :: Basic/./SwiftBasicTests/Cache.sameKeyValueDestroysKey' FAILED ********************
Note: Google Test filter = Cache.sameKeyValueDestroysKey
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from Cache
[ RUN      ] Cache.sameKeyValueDestroysKey
=================================================================
==72020==ERROR: AddressSanitizer: stack-use-after-scope on address 0x7ffee35322b4 at pc 0x00010c7cfa87 bp 0x7ffee3532110 sp 0x7ffee3532108
READ of size 4 at 0x7ffee35322b4 thread T0
    #0 0x10c7cfa86 in swift::sys::Cache<(anonymous namespace)::KeyCounter, (anonymous namespace)::Counter, swift::sys::CacheKeyInfo<(anonymous namespace)::KeyCounter>, swift::sys::CacheValueInfo<(anonymous namespace)::Counter> >::valueRelease(void*, void*) (SwiftBasicTests:x86_64+0x100102a86)
    #1 0x7fff77ca673a in _value_entry_remove (libcache.dylib:x86_64+0x273a)
    #2 0x7fff77ca65f0 in _entry_remove (libcache.dylib:x86_64+0x25f0)
    #3 0x7fff77ca6979 in _cache_remove_all_locked (libcache.dylib:x86_64+0x2979)
    #4 0x7fff77ca5839 in cache_release (libcache.dylib:x86_64+0x1839)
    #5 0x7fff77ca69a8 in cache_destroy (libcache.dylib:x86_64+0x29a8)
    #6 0x10c7c8ca6 in Cache_sameKeyValueDestroysKey_Test::TestBody() (SwiftBasicTests:x86_64+0x1000fbca6)
    #7 0x10ca6c1e7 in testing::Test::Run() gtest.cc
    #8 0x10ca6e712 in testing::TestInfo::Run() gtest.cc:2656
    #9 0x10ca6fbb6 in testing::TestCase::Run() gtest.cc:2774
    #10 0x10ca8cf96 in testing::internal::UnitTestImpl::RunAllTests() gtest.cc:4649
    #11 0x10ca8c4f7 in testing::UnitTest::Run() gtest.cc
    #12 0x10ca53ea0 in main gtest.h:2233
    #13 0x7fff77e213d4 in start (libdyld.dylib:x86_64+0x163d4)

Address 0x7ffee35322b4 is located in stack of thread T0 at offset 212 in frame
    #0 0x10c7c7eff in Cache_sameKeyValueDestroysKey_Test::TestBody() (SwiftBasicTests:x86_64+0x1000faeff)

  This frame has 21 object(s):
    [32, 80) 'CBs.i.i'
    [112, 120) 'cache'
    [144, 156) 'k1'
    [176, 188) 'k2'
    [208, 216) 'c' <== Memory access at offset 212 is inside this variable
    [240, 256) 'gtest_ar'
    [272, 276) 'ref.tmp'
    [288, 296) 'ref.tmp2'
    [320, 328) 'ref.tmp3'
    [352, 368) 'gtest_ar5'
    [384, 388) 'ref.tmp6'
    [400, 408) 'ref.tmp10'
    [432, 440) 'ref.tmp11'
    [464, 480) 'gtest_ar14'
    [496, 500) 'ref.tmp15'
    [512, 520) 'ref.tmp20'
    [544, 552) 'ref.tmp21'
    [576, 592) 'gtest_ar24'
    [608, 612) 'ref.tmp25'
    [624, 632) 'ref.tmp30'
    [656, 664) 'ref.tmp31'
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-use-after-scope (SwiftBasicTests:x86_64+0x100102a86) in swift::sys::Cache<(anonymous namespace)::KeyCounter, (anonymous namespace)::Counter, swift::sys::CacheKeyInfo<(anonymous namespace)::KeyCounter>, swift::sys::CacheValueInfo<(anonymous namespace)::Counter> >::valueRelease(void*, void*)
Shadow bytes around the buggy address:
  0x1fffdc6a6400: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1fffdc6a6410: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1fffdc6a6420: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1fffdc6a6430: 00 00 00 00 00 00 00 00 00 00 00 00 f1 f1 f1 f1
  0x1fffdc6a6440: f8 f8 f8 f8 f8 f8 f2 f2 f2 f2 00 f2 f2 f2 f8 f8
=>0x1fffdc6a6450: f2 f2 f8 f8 f2 f2[f8]f2 f2 f2 f8 f8 f2 f2 f8 f2
  0x1fffdc6a6460: f8 f2 f2 f2 f8 f2 f2 f2 f8 f8 f2 f2 f8 f2 f8 f2
  0x1fffdc6a6470: f2 f2 f8 f2 f2 f2 f8 f8 f2 f2 f8 f2 f8 f2 f2 f2
  0x1fffdc6a6480: f8 f2 f2 f2 f8 f8 f2 f2 f8 f2 f8 f2 f2 f2 f8 f3
  0x1fffdc6a6490: f3 f3 f3 f3 00 00 00 00 00 00 00 00 00 00 00 00
  0x1fffdc6a64a0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==72020==ABORTING
```

**Comment by marcrasi** [comment]
Yes, that's happening in the test I added. I think I see a fix. I'll create a PR.

**Comment by marcrasi** [comment]
This should be fixed now: https://github.com/apple/swift/pull/26225
